src/compare.c

```diff
@@ -122,11 +122,13 @@ int compare_scalar(SEXP x, R_len_t i, SEXP y, R_len_t j, bool na_equal) {
 // R interface -----------------------------------------------------------------
 
 SEXP vctrs_compare(SEXP x, SEXP y, SEXP na_compare_) {
-  if (TYPEOF(x) != TYPEOF(y) || vec_size(x) != vec_size(y))
+  R_len_t n = vec_size(x);
+
+  if (TYPEOF(x) != TYPEOF(y) || n != vec_size(y)) {
     Rf_errorcall(R_NilValue, "`x` and `y` must have same types and lengths");
```

Comment:
Since we recycle to common size and cast to common type at the R level, is this error even reachable? Maybe it's good practice to keep it in, just something I noticed.

Reply:
I was thinking we should probably remove these checks.